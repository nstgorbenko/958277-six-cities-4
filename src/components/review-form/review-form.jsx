import {connect} from "react-redux";
import React, {Fragment, PureComponent} from "react";
import PropTypes from "prop-types";

import {getLoadStatus} from "../../reducer/data/selectors.js";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {LoadStatus} from "../../reducer/data/data.js";

const RATING_TITLES = [`perfect`, `good`, `not bad`, `badly`, `terribly`];

class ReviewForm extends PureComponent {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(evt) {
    const {id, rating, review, loadStatus, onSubmit, onReset} = this.props;

    evt.preventDefault();
    onSubmit({
      hotelId: id,
      comment: review,
      rating,
    })
    .then(() => {
      if (loadStatus === LoadStatus.SUCCESS) {
        onReset();
      }
    });
  }

  render() {
    const {rating, review, loadStatus, onChange} = this.props;
    const isDisabledInput = loadStatus === LoadStatus.LOADING;
    const isDisabledButton = rating === 0 || review.length < 50 || review.length > 300 || isDisabledInput;

    return (
      <form className="reviews__form form" action="#" method="post"
        onSubmit={this.handleSubmit}
      >
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <div className="reviews__rating-form form__rating">
          {RATING_TITLES.map((ratingTitle, index) => {
            const id = RATING_TITLES.length - index;
            return (
              <Fragment key={ratingTitle}>
                <input className="form__rating-input visually-hidden" name="rating" type="radio"
                  value={id}
                  id={`${id}-stars`}
                  checked={id === rating}
                  disabled={`${isDisabledInput ? `disabled` : ``}`}
                  onChange={onChange}
                />
                <label className="reviews__rating-label form__rating-label"
                  htmlFor={`${id}-stars`}
                  title={ratingTitle}
                >
                  <svg className="form__star-image" width="37" height="33">
                    <use xlinkHref="#icon-star"></use>
                  </svg>
                </label>
              </Fragment>
            );
          })}
        </div>
        <textarea className="reviews__textarea form__textarea" id="review" name="review" placeholder="Tell how was your stay, what you like and what can be improved"
          minLength="50"
          maxLength="300"
          value={review}
          onChange={onChange}
          disabled={`${isDisabledInput ? `disabled` : ``}`}
        ></textarea>
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and describe your stay with at least <b className="reviews__text-amount">50 characters</b>.
          </p>
          <button className="reviews__submit form__submit button" type="submit" disabled={`${isDisabledButton ? `disabled` : ``}`}>Submit</button>
        </div>
        {loadStatus === LoadStatus.ERROR && <p className="reviews__text-amount">Something went wrong. Please try again later</p>}
      </form>
    );
  }
}

ReviewForm.propTypes = {
  id: PropTypes.number.isRequired,
  rating: PropTypes.number.isRequired,
  review: PropTypes.string.isRequired,
  loadStatus: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onReset: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  loadStatus: getLoadStatus(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onSubmit(reviewData) {
    return dispatch(DataOperation.postReview(reviewData));
  }
});

export {ReviewForm};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewForm);
