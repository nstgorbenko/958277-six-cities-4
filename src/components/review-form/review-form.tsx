import {connect} from "react-redux";
import * as React from 'react';

import {getLoadStatus} from "../../reducer/data/selectors";
import {Operation as DataOperation} from "../../reducer/data/data";
import {LoadStatus} from "../../reducer/data/data";
import {RATING_TITLES, ReviewLength} from "../../const";

interface Props {
  id: number;
  rating: number;
  review: string;
  loadStatus: string;
  onChange(): void;
  onReset(): void;
  onSubmit(reviewData: {
    hotelId: number;
    comment: string;
    rating: number;
  }): Promise<void>;
}

class ReviewForm extends React.PureComponent<Props> {
  props: Props;

  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(evt) {
    const {id, rating, review, loadStatus, onSubmit, onReset} = this.props;

    evt.preventDefault();
    onSubmit({
      hotelId: id,
      comment: review,
      rating,
    })
    .then(() => {
      if (loadStatus === LoadStatus.SUCCESS) {
        onReset();
      }
    });
  }

  render() {
    const {rating, review, loadStatus, onChange} = this.props;
    const isDisabledInput: boolean = loadStatus === LoadStatus.LOADING;
    const isDisabledButton: boolean = rating === 0 || review.length < ReviewLength.MIN || review.length > ReviewLength.MAX || isDisabledInput;

    return (
      <form className="reviews__form form" action="#" method="post"
        onSubmit={this.handleSubmit}
      >
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <div className="reviews__rating-form form__rating">
          {RATING_TITLES.map((ratingTitle, index) => {
            const id: number = RATING_TITLES.length - index;
            return (
              <React.Fragment key={ratingTitle}>
                <input className="form__rating-input visually-hidden" name="rating" type="radio"
                  value={id}
                  id={`${id}-stars`}
                  checked={id === rating}
                  disabled={isDisabledInput}
                  onChange={onChange}
                />
                <label className="reviews__rating-label form__rating-label"
                  htmlFor={`${id}-stars`}
                  title={ratingTitle}
                >
                  <svg className="form__star-image" width="37" height="33">
                    <use xlinkHref="#icon-star"></use>
                  </svg>
                </label>
              </React.Fragment>
            );
          })}
        </div>
        <textarea className="reviews__textarea form__textarea" id="review" name="review" placeholder="Tell how was your stay, what you like and what can be improved"
          minLength={ReviewLength.MIN}
          maxLength={ReviewLength.MAX}
          value={review}
          onChange={onChange}
          disabled={isDisabledInput}
        ></textarea>
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and describe your stay with at least <b className="reviews__text-amount">50 characters</b>.
          </p>
          <button className="reviews__submit form__submit button" type="submit" disabled={isDisabledButton}>Submit</button>
        </div>
        {loadStatus === LoadStatus.ERROR && <p className="reviews__text-amount error-message">Something went wrong. Please try again later</p>}
      </form>
    );
  }
}

const mapStateToProps = (state) => ({
  loadStatus: getLoadStatus(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onSubmit(reviewData) {
    return dispatch(DataOperation.postReview(reviewData));
  }
});

export {ReviewForm};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewForm);
