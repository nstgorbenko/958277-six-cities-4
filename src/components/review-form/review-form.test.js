import React from "react";
import renderer from "react-test-renderer";

import {ReviewForm} from "./review-form.jsx";

describe(`ReviewForm Component rendering`, () => {
  it(`renders correctly with loadStatus equal SUCCESS`, () => {
    const tree = renderer
      .create(<ReviewForm
        id={1}
        rating={5}
        review={``}
        loadStatus={`SUCCESS`}
        onChange={() => {}}
        onReset={() => {}}
        onSubmit={() => {}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`renders correctly with loadStatus equal LOADING`, () => {
    const tree = renderer
      .create(<ReviewForm
        id={1}
        rating={5}
        review={``}
        loadStatus={`LOADING`}
        onChange={() => {}}
        onReset={() => {}}
        onSubmit={() => {}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`renders correctly with loadStatus equal ERROR`, () => {
    const tree = renderer
      .create(<ReviewForm
        id={1}
        rating={5}
        review={``}
        loadStatus={`ERROR`}
        onChange={() => {}}
        onReset={() => {}}
        onSubmit={() => {}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`renders correctly with loadStatus equal SUCCESS and set rating and review values`, () => {
    const tree = renderer
      .create(<ReviewForm
        id={1}
        rating={5}
        review={`More than 50 characters of my review about staying in this hotel.`}
        loadStatus={`SUCCESS`}
        onChange={() => {}}
        onReset={() => {}}
        onSubmit={() => {}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
