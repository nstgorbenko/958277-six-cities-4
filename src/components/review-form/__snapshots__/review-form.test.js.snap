// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm Component rendering renders correctly with loadStatus equal ERROR 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={true}
      className="form__rating-input visually-hidden"
      disabled=""
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled=""
    id="review"
    maxLength={300}
    minLength={50}
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled="disabled"
      type="submit"
    >
      Submit
    </button>
  </div>
  <p
    className="reviews__text-amount"
  >
    Something went wrong. Please try again later
  </p>
</form>
`;

exports[`ReviewForm Component rendering renders correctly with loadStatus equal LOADING 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={true}
      className="form__rating-input visually-hidden"
      disabled="disabled"
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled="disabled"
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled="disabled"
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled="disabled"
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled="disabled"
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled="disabled"
    id="review"
    maxLength={300}
    minLength={50}
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled="disabled"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`ReviewForm Component rendering renders correctly with loadStatus equal SUCCESS 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={true}
      className="form__rating-input visually-hidden"
      disabled=""
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled=""
    id="review"
    maxLength={300}
    minLength={50}
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled="disabled"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`ReviewForm Component rendering renders correctly with loadStatus equal SUCCESS and set rating and review values 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={true}
      className="form__rating-input visually-hidden"
      disabled=""
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      disabled=""
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    disabled=""
    id="review"
    maxLength={300}
    minLength={50}
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value="More than 50 characters of my review about staying in this hotel."
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
